editor/src/messages/portfolio/document/properties_panel/utility_functions.rs

```diff
@@ -1163,60 +1166,218 @@ fn node_gradient_type(gradient: &Gradient) -> LayoutGroup {
-			})),
-		],
+			})));
+		}
+		widgets.push(WidgetHolder::new(Widget::IconButton(IconButton {
+			icon: "Add".to_string(),
+			tooltip: "Add a gradient stop after this".to_string(),
+			size: 16,
+			on_update,
+			..Default::default()
+		})));
 	}
+	LayoutGroup::Row { widgets }
 }
 
-fn node_section_fill(fill: &Fill) -> Option<LayoutGroup> {
+fn node_section_fill(fill: &Fill, layer_bounds: Option<[DVec2; 2]>) -> Option<LayoutGroup> {
+	let initial_color = if let Fill::Solid(color) = fill { *color } else { Color::BLACK };
+
 	match fill {
 		Fill::Solid(_) | Fill::None => Some(LayoutGroup::Section {
 			name: "Fill".into(),
-			layout: vec![LayoutGroup::Row {
-				widgets: vec![
-					WidgetHolder::new(Widget::TextLabel(TextLabel {
-						value: "Color".into(),
-						..TextLabel::default()
-					})),
-					WidgetHolder::new(Widget::Separator(Separator {
-						separator_type: SeparatorType::Unrelated,
-						direction: SeparatorDirection::Horizontal,
-					})),
-					WidgetHolder::new(Widget::ColorInput(ColorInput {
-						value: if let Fill::Solid(color) = fill { Some(*color) } else { None },
-						on_update: WidgetCallback::new(|text_input: &ColorInput| {
-							let fill = if let Some(value) = text_input.value { Fill::Solid(value) } else { Fill::None };
-							PropertiesPanelMessage::ModifyFill { fill }.into()
-						}),
-						..ColorInput::default()
-					})),
-				],
-			}],
+			layout: {
+				let mut result = vec![LayoutGroup::Row {
+					widgets: vec![
+						WidgetHolder::new(Widget::TextLabel(TextLabel {
+							value: "Color".into(),
+							..TextLabel::default()
+						})),
+						WidgetHolder::new(Widget::Separator(Separator {
+							separator_type: SeparatorType::Unrelated,
+							direction: SeparatorDirection::Horizontal,
+						})),
+						WidgetHolder::new(Widget::ColorInput(ColorInput {
+							value: if let Fill::Solid(color) = fill { Some(*color) } else { None },
+							on_update: WidgetCallback::new(|text_input: &ColorInput| {
+								let fill = if let Some(value) = text_input.value { Fill::Solid(value) } else { Fill::None };
+								PropertiesPanelMessage::ModifyFill { fill }.into()
+							}),
+							..ColorInput::default()
+						})),
+					],
+				}];
+				if let Some(bounds) = layer_bounds {
+					result.push(LayoutGroup::Row {
+						widgets: vec![
+							WidgetHolder::new(Widget::TextLabel(TextLabel {
+								value: "".into(),
+								..TextLabel::default()
+							})),
+							WidgetHolder::new(Widget::Separator(Separator {
+								separator_type: SeparatorType::Unrelated,
+								direction: SeparatorDirection::Horizontal,
+							})),
+							WidgetHolder::new(Widget::TextButton(TextButton {
+								label: "Use Gradient".into(),
+								tooltip: "Change this fill from a solid color to a gradient".into(),
+								on_update: WidgetCallback::new(move |_: &TextButton| {
+									// TODO: Fix bounds calculation for shapes with a 1x1 scale (text, Pen tool shape, other non-primitives)
```

Comment:
I need help fixing this bug since I don't understand our API surface for transforms and bounds well enough to make it work in the case described here. I'd appreciate if you could help with that before merging @0HyperCube. But we're otherwise good to go, everything is working nicely! Thanks 🙏.